### PR TITLE
Logout

### PR DESCRIPTION
**Pull request by ibeckermayer**
Adds logout functionality

## backend
Adds a goroutine-safe method `SessionManager.DeleteSession()` for deleting a session from the session manager, and the logout handler that calls it. For edge cases where the "api/logout" request is sent with a missing/improperly-formatted Authorization header, or the token sent in the Authorization header is not found in the `SessionManager`, I've elected to simply log the discrepancy and return a 204 success response to the client so that it still reacts the same as a normal logout (where a session of sessionID is deleted). Open to better ideas/security concerns.

## frontend
The frontend route structure is revamped to direct the user directly to the "/dashboard" route, which is now protected by an `<Authenticated>` wrapper. 

Note: In `<Authenticated>` (and elsewhere), I've elected to use conditional rendering with `<Redirect>` over the `useEffect` hook with `history.push`; unlike the `useEffect` approach, conditional rendering prevents the app from flashing the primary component on the screen before redirecting the user (e.g. see changes starting on lines 25/26 of `Login.jsx`)

## To test
To check manually that the `SessionManager` is behaving as expected, it can be helpful to add `log.Printf("%v", lh.sm)` to line 73 of `login.go` and line 34 of `logout.go`

Run
```bash
docker-compose -f docker-compose-dev.yml up
```
and navigate to https://0.0.0.0:8080

Try navigating directly to each of
* https://0.0.0.0:8080
* https://0.0.0.0:8080/login
* https://0.0.0.0:8080/dashboard

while logged out, and see that they behave as expected. Login with username/pwd = `dev@goteleport.com` and check that you're redirected to the dashboard. 

Try each of the links above again and see that they behave as expected while logged in. 

Hit the "Logout" button and see that you're redirected to /login. Confirm that the `SessionManager` is behaving as it should via the log outputs.

### CODE REVIEW

File: backend/internal/handlers/logout.go
Comment (alex-kovoy) [inline backend/internal/handlers/logout.go]: Lets introduce a wrapper function and use it to wrap protected handlers. Something like below 

```golang
func withAuth(handler ProtectedHandler) httprouter.Handle {
    return func(w http.ResponseWriter, r *http.Request) {
        session, err := authenticate(w, r)
        if err != nil  {
            http.Error(w, "Forbidden", http.StatusForbidden)
            return
        }

        // Pass a session 
        handler.ServeHTTP(w, r, session)
    }
}

// and then
srv.router.Handle("/api/logout", withAuth(logoutHandler)).Methods("DELETE")
srv.router.Handle("/api/plan", withAuth(upgradePlanHandler)).Methods("PUT")
```

Comment (alex-kovoy) [reply backend/internal/handlers/logout.go]: (removed the pseudocode from the sentence as it does not need to be more complicated than this)

Comment (ibeckermayer) [reply backend/internal/handlers/logout.go]: Good idea. I was digging into the [scs](https://github.com/alexedwards/scs) source to figure out how he made the generalizable middlewear (he uses Context), but decided that would be overkill for this project. This is much simpler though.

Comment (alex-kovoy) [reply backend/internal/handlers/logout.go]: Using request.Context is actually a good idea to pass a userId parameter. This way you do not need to have a custom handler methods to pass the session.

```golang

func withAuth(handler http.Handler) httprouter.Handle {
    return func(w http.ResponseWriter, r *http.Request) {
        userId, err := authenticate(w, r)
        if err != nil  {
            http.Error(w, "Forbidden", http.StatusForbidden)
            return
        }
 
        r = r.WithContext( // add userId to the context)
        handler.ServeHTTP(w, r)
    }
}
```

**Review by alex-kovoy — APPROVED** [review]

**Comment by alex-kovoy** [comment]
@ibeckermayer lets wait for 2 approved reviews next time before merging.  @awly could you take a look at this one and @ibeckermayer will address your feedback in the next PR?

**Comment by ibeckermayer** [comment]
> @ibeckermayer lets wait for 2 approved reviews next time before merging. @awly could you take a look at this one and @ibeckermayer will address your feedback in the next PR?

Oh my bad, I misinterpreted our Slack convo. I'll create a new branch for making use of context and whatever feedback @awly has.

File: backend/internal/auth/sessionmanager.go
Comment (awly) [inline backend/internal/auth/sessionmanager.go:88]: nit: all of this should happen under a single write lock
if you have 2 concurrent `DeleteSession` calls, they might intertwine their read/write locks and return the wrong bool.

File: backend/internal/handlers/logout.go
Comment (awly) [inline backend/internal/handlers/logout.go:32]: should logout redirect to the login page?

File: backend/internal/handlers/utils.go
Comment (awly) [inline backend/internal/handlers/utils.go:84]: nit: this check should be `splitToken := strings.Fields(reqToken); if len(splitToken) != 2 || splitToken[0] != "Bearer" { ...`, to avoid other invalid values like `Bearer foo bar` for example

**Review by awly — COMMENTED** [review]
Left some small comments, feel free to bundle them in a future PR

File: backend/internal/auth/sessionmanager.go
Comment (ibeckermayer) [reply backend/internal/auth/sessionmanager.go:88]: Good catch

File: backend/internal/handlers/logout.go
Comment (ibeckermayer) [reply backend/internal/handlers/logout.go:32]: Yes, but for this app all routing is handled client side. In this case the client is wired up so that when it receives this response, it deletes the session data from localStorage which triggers it to flip to the "/login" screen.

**Pull request merged**
